Transition plan for modularization (#701)
kyma-project/community: concepts/modularization/transition.md

# Transition from reconciler to kyma-operator

With almost 20 modules and several teams involved it is not feasible to coordinate the transition to the new modular architecture in the single release. The plan is to switch modules to the new architecture gradually, avoiding big bang releases and keeping existing clusters stable. 

## Phase 1 - local development

> You can start developing module operator with the provided guide, tools and libraries. Following the guide you will generate the operator scaffold, copy your helm chart inside and use the [`manifest` package](https://github.com/kyma-project/manifest-operator#use-manifest-packages-in-your-own-operator) to install it in your reconciliation logic. You can add configuration options to the spec of your operator CRD and use it as overrides for helm charts or in your custom logic (you still can use custom code in the operator).

Reviewer: ```suggestion
You can start developing a module operator with the provided guide, tools and libraries. Following the guide, you will generate the module and operator scaffold, copy your helm chart inside and use the [`manifest` package](https://github.com/kyma-project/manifest-operator#use-manifest-packages-in-your-own-operator) to install it in your reconciliation logic. You can add configuration options to the spec of your operator CRD and use it as overrides for helm charts or in your custom logic (you still can use custom code in the operator). This also allows you to differentiate different installation profiles and code paths.
```

Reviewer 2: following up on Jakob's comment:

```suggestion
You can start developing a module operator with the provided guide, tools, and libraries. Following the guide, you generate the module and operator scaffold, copy your Helm chart inside, and use the [`manifest` package](https://github.com/kyma-project/manifest-operator#use-manifest-packages-in-your-own-operator) to install it in your reconciliation logic. You can add configuration options to the spec of your operator CRD and use them as overrides for Helm charts or in your custom logic (you still can use custom code in the operator). This way, you can also differentiate several installation profiles and code paths.
```

I don't love "allow" for anything that wasn't forbidden before. Usually it's  better to say "enables you to", or, simpler "you can".

> You can test operator on its own by applying operator CRD and creating the instance (you can install operator on your test cluster or even run it locally). When your operator works you can build the image and publish it (using the recommended pipeline template) and create a `ModuleTemplate` that contains a default custom resource used to install the module and operator deployment. You create one module template per release channel you want to support. Having `ModuleTemplate` you can start testing your module with `kyma-operator` . To do that you need to install kyma-operator and its custom resources in your test cluster, create a `Kyma` resource and add your module in the spec. `Kyma-operator` will install your operator CRD, will deploy your operator and start module installation by creating default custom resource. When your operator changes the status.state field it should be propagated to the `Kyma` resource status. You can install `kyma-operator` and enable/disable modules using `kyma CLI` or `kyma Dashboard`.

Reviewer: ```suggestion
You can test the operator on its own by applying the operator CRD and starting it independently (you can install operator on your test cluster or even run it locally). You can build the operator image and publish it (using the recommended pipeline template) and generate a `ModuleTemplate` that contains a default custom resource used to install the module and operator deployment. You create one `ModuleTemplate` per release channel you want to support. 

With the `ModuleTemplate`, you can integrate your module with `kyma-operator` . To do that, you need to install kyma-operator and its custom resources in your test cluster, create a `Kyma` resource and add your module in the spec. `Kyma-operator` will install your operator CRD, will deploy your operator and start the module installation by creating the default custom resource. When your operator changes the status.state field (for an example, take a look at the [template-operator](https://github.com/kyma-project/kyma-operator/blob/main/samples/template-operator/api/v1alpha1/mockup_types.go#L38-L54)) it should be propagated to the `Kyma` resource status. You can install `kyma-operator` and will be able to enable/disable modules using `kyma CLI` or `kyma Dashboard`.
```

## Phase 2 - first module managed by `kyma-operator` integrated with KEB

### Phase 2a - new clusters only (internal usage)

`kyma-operator` is deployed in the control-plane and integrated with Kyma Environment Broker (KEB). Additional plan (preview) is created in KEB that creates Kyma clusters with kyma-operator only (no call to reconciler). This plan is the first playground to test KEB integration without affecting existing clusters and flows. The plan is tested with all environments (dev, stage, prod) to ensure we have all major integration issues resolved. The first module (frontrunner) can be submitted by the team and installed in the cluster. The minimal submission process is described and fully automated (manual steps are allowed only before submission). The minimal submission process looks like:

> - team verifies on their own if new module (or version) is product standard compliant 

Reviewer: how do they do this?

Author: We do not imply how to do that in the modularization concept. It is just a clear statement that it is the team's responsibility to deliver modules with proper quality. We can't validate it.

> - team creates pull request to the control-plane that contains new/updated ModuleTemplate (by doing it team declares product standard compliance and functional correctness - can be checkbox in PR template)

Reviewer: ```suggestion
- team creates pull request to the control-plane that contains new/updated ModuleTemplate(s) (by doing it team declares product standard compliance and functional correctness - can be checkbox in PR template)
```
we have 3 ModuleTemplates for the different release channels

- automated test verifies if the module operator can be installed/upgraded without issues for all release channels (with default settings) - it is not functional test or integration test 

> - automated test verifies if all images installed by operator are declared in the module template, signed by approved CA, vulnerability free

Reviewer: ```suggestion
- automated test verifies if all images installed by operator are declared in the module template, signed by approved CA, vulnerability free
```
I checked with @clebs and we want to check if it makes sense bundling the ModuleTemplate Information into the existing built module. That way we have only one image and all the checks can be done at build time so we could skip this

Author: We have to discuss it. Example:
Eventing operator installs eventing-controller and eventing-publishing-proxy. These 2 components are not directly installed by the kyma-operator, but should be signed and vulnerability free.

Reviewer 2: ```suggestion
4. An automated test verifies if all images installed by operator are declared in the module template, signed by approved CA, and free of vulnerabilities.
```

# Transition from reconciler to kyma-operator

> With almost 20 modules and several teams involved it is not feasible to coordinate the transition to the new modular architecture in the single release. The plan is to switch modules to the new architecture gradually, avoiding big bang releases and keeping existing clusters stable. 

Reviewer 2: ```suggestion
With almost 20 modules and several teams involved, it is not feasible to coordinate the transition to the new modular architecture in a single release. The plan is to switch modules to the new architecture gradually, avoiding big bang releases and keeping existing clusters stable. 
```

## Phase 1 - local development

You can start developing module operator with the provided guide, tools and libraries. Following the guide you will generate the operator scaffold, copy your helm chart inside and use the [`manifest` package](https://github.com/kyma-project/manifest-operator#use-manifest-packages-in-your-own-operator) to install it in your reconciliation logic. You can add configuration options to the spec of your operator CRD and use it as overrides for helm charts or in your custom logic (you still can use custom code in the operator).

> You can test the operator on its own by applying the operator CRD and starting it independently (you can install operator on your test cluster or even run it locally). You can build the operator image and publish it (using the recommended pipeline template) and generate a `ModuleTemplate` that contains a default custom resource used to install the module and operator deployment. You create one `ModuleTemplate` per release channel you want to support. 

Reviewer 2: ```suggestion
To test the operator on its own, apply the operator CRD and start it independently. You can install the operator on your test cluster or even run it locally.  
You can build the operator image and publish it (using the recommended pipeline template), and generate a `ModuleTemplate` that contains a default custom resource used to install the module and operator deployment. For every release channel you want to support, create one `ModuleTemplate`. 
```

> With the `ModuleTemplate`, you can integrate your module with `kyma-operator` . To do that, you need to install kyma-operator and its custom resources in your test cluster, create a `Kyma` resource and add your module in the spec. `Kyma-operator` will install your operator CRD, will deploy your operator and start the module installation by creating the default custom resource. When your operator changes the status.state field (for an example, take a look at the [template-operator](https://github.com/kyma-project/kyma-operator/blob/main/samples/template-operator/api/v1alpha1/mockup_types.go#L38-L54)) it should be propagated to the `Kyma` resource status. You can install `kyma-operator` and will be able to enable/disable modules using `kyma CLI` or `kyma Dashboard`.

Reviewer 2: ```suggestion
With the `ModuleTemplate`, you can integrate your module with `kyma-operator`: Install kyma-operator and its custom resources in your test cluster, create a `Kyma` resource, and add your module in the spec. `Kyma-operator` installs your operator CRD, deploys your operator, and starts the module installation by creating the default custom resource. When your operator changes the status.state field (for an example, take a look at the [template-operator](https://github.com/kyma-project/kyma-operator/blob/main/samples/template-operator/api/v1alpha1/mockup_types.go#L38-L54)), it should be propagated to the `Kyma` resource status. You can install `kyma-operator` and will be able to enable and disable modules using `Kyma CLI` or `Kyma Dashboard`.
```

What does "should" mean here? Is it a recommendation? Are there other options (then use "can")? Is it mandatory/necessary (then use "must))? What if it doesn't propagate the status information? Could we simply drop the "should" and write "..., it is propagated to the `Kyma` resource status."

## Phase 2 - first module managed by `kyma-operator` integrated with KEB

### Phase 2a - new clusters only (internal usage)

> `kyma-operator` is deployed in the control-plane and integrated with Kyma Environment Broker (KEB). Additional plan (preview) is created in KEB that creates Kyma clusters with kyma-operator only (no call to reconciler). This plan is the first playground to test KEB integration without affecting existing clusters and flows. The plan is tested with all environments (dev, stage, prod) to ensure we have all major integration issues resolved. The first module (frontrunner) can be submitted by the team and installed in the cluster. The minimal submission process is described and fully automated (manual steps are allowed only before submission). The minimal submission process looks like:

Reviewer 2: ```suggestion
`kyma-operator` is deployed in the control-plane and integrated with Kyma Environment Broker (KEB). As a first playground to test KEB integration without affecting existing clusters and flows, an additional plan called "preview" is created in KEB. This plan creates Kyma clusters with kyma-operator only and does not call reconciler.
The plan is tested with all environments (dev, stage, prod) to ensure we have all major integration issues resolved. The first module (frontrunner) can be submitted by the team and installed in the cluster. The minimal submission process is described and fully automated (manual steps are allowed only before submission). The minimal submission process looks like:
```

> - team verifies on their own if new module (or version) is product standard compliant 

Reviewer 2: I guess the order of the steps matters, so we should use a numbered list.

```suggestion
1. The team verifies on their own if the new module (or version) complies with the SAP Product Standard.
```

Can we link to the relevant SAP Product Standard(s) here? Or do you refer to other product standards?

> - team creates pull request to the control-plane that contains new/updated ModuleTemplate(s) (by doing it team declares product standard compliance and functional correctness - can be checkbox in PR template)

Reviewer 2: ```suggestion
2. The team creates a pull request to the control-plane that contains one or more new or updated ModuleTemplates. With this action, the team declares product standard compliance and functional correctness; for example, with a checkbox in the PR template.
```

> - automated test verifies if the module operator can be installed/upgraded without issues for all release channels (with default settings) - it is not functional test or integration test 

Reviewer 2: ```suggestion
3. An automated test verifies if the module operator can be installed or upgraded without issues for all release channels (with default settings). This is not a functional test or integration test!
```

- automated test verifies if all images installed by operator are declared in the module template, signed by approved CA, vulnerability free

> - control-plane maintainer (codeowner) approves the PR which is automatically merged and rolled out

Reviewer 2: ```suggestion
5. The control-plane maintainer (code owner) approves the PR, which is automatically merged and rolled out.
```

### Phase 2b - move first module (frontrunner) from reconciler to kyma-operator (production)

> Regular plans use reconciler and kyma-operator in parallel (clusters are managed by both components, but their lists of modules don't overlap). Frontrunner module is disabled in the reconciler (ignored / replaced by dummy reconciler). `Kyma` resources containing the first module are populated for all existing clusters using script or KEB functionality. In each cluster `kyma-operator` installs frontrunner module operator with its CRD (configuration specification), and creates default configuration resource for the module. Module operator installs all needed components (including UI config map for Kyma Dashboard) and updated status of configuration custom resource (status.state = Ready). From that point in time cluster admin (or any user with sufficient roles) can edit the configuration resource to change module settings or can remove the module completely by deleting it from the list in the `Kyma` resource.

Reviewer 2: Are "frontrunner" and "first" synonyms here? If yes, I suggest to keep just one term for better consistency.

```suggestion
Regular plans use reconciler and kyma-operator in parallel (clusters are managed by both components, but their lists of modules don't overlap). The frontrunner module is disabled in the reconciler (ignored or replaced by a dummy reconciler). `Kyma` resources containing the frontrunner module are populated for all existing clusters using a script or KEB functionality. In each cluster, `kyma-operator` installs the frontrunner module operator with its CRD (configuration specification), and creates a default configuration resource for the module. Module operator installs all needed components (including UI config map for Kyma Dashboard) and updates the status of the configuration custom resource (status.state = Ready). From that point in time, the cluster admin (or any user with sufficient authorization) can edit the configuration resource to change the module settings, or can remove the module completely by deleting it from the list in the `Kyma` resource.
```

I'm not sure I got the "updated" vs "updates" suggestion correctly; this sentence was a bit confusing.

The submission process is extended with all aspects required for production deployment, like support, commercialization, security, micro-delivery, etc.
## Phase 3 - all modules are managed by `kyma-operator` 

> All modules are enabled in `kyma-operator` and modules are added to Kyma resources for existing clusters. Reconciler is shutdown and KEB plans integrate only with `kyma-operator` (preview plan is also gone). All teams have submitted their modules using full version of submission process - the minimal submission process introduced in phase 2 extended to cover other aspects like support, microdelivery, etc. Other SAP teams can also submit their modules following improved technical guides and submission process description. 

Reviewer 2: ```suggestion
All modules are enabled in `kyma-operator`, and modules are added to Kyma resources for existing clusters. Reconciler is shut down, and KEB plans integrate only with `kyma-operator`. The "preview" plan is also gone. All teams have submitted their modules following the full submission process - the minimal submission process introduced in phase 2 extended to cover other aspects like support, microdelivery, etc. Other SAP teams can also submit their modules following improved technical guides and the full submission process. 
```
